### DISCUSSION TITLE
Object and BP layers

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#61] [Q&A] [answered]
I would like to understand the concept of layers a bit better. From what the description says, I understand that object layer is a separate BP quad tree. This is pretty intuitive - I define a group that can collide or not with another group. But I am not sure what a BP layer means and how it is related to the object layer:
https://jrouwe.github.io/JoltPhysics/index.html

We also have an example:
```js
// We use only 2 layers: one for non-moving objects and one for moving objects
let object_filter = new Jolt.ObjectLayerPairFilterTable(2);
object_filter.EnableCollision(LAYER_NON_MOVING, LAYER_MOVING);
object_filter.EnableCollision(LAYER_MOVING, LAYER_MOVING);

// We use a 1-to-1 mapping between object layers and broadphase layers
const BP_LAYER_NON_MOVING = new Jolt.BroadPhaseLayer(0);
const BP_LAYER_MOVING = new Jolt.BroadPhaseLayer(1);
let bp_interface = new Jolt.BroadPhaseLayerInterfaceTable(2, 2);
bp_interface.MapObjectToBroadPhaseLayer(LAYER_NON_MOVING, BP_LAYER_NON_MOVING);
bp_interface.MapObjectToBroadPhaseLayer(LAYER_MOVING, BP_LAYER_MOVING);
```

I am not sure why we need BP layers, or how to map a one-to-many relationship. What would be an example use case for one-to-one or one-to-many relationships? Apart from use-case would this syntax be correct?:

```js
let bp_interface = new Jolt.BroadPhaseLayerInterfaceTable(2, 3);
bp_interface.MapObjectToBroadPhaseLayer(LAYER_NON_MOVING, BP_LAYER_NON_MOVING);
bp_interface.MapObjectToBroadPhaseLayer(LAYER_MOVING, BP_LAYER_MOVING);
bp_interface.MapObjectToBroadPhaseLayer(LAYER_MOVING, BP_LAYER_MOVING_SUPER_FAST);
```

Do we want to prefer having less object layers over more BP layers?

**Comment by LeXXik** [comment]
On the same note, bodies belong to collision groups. If I understand correctly, then these groups specify if a body hits another body within the same object layer. If true, then what are sub-groups? Can a body belong to more than one group/sub-group?

**Comment by jrouwe** [comment, marked as answer]
Hello,

I've added a bit more documentation to the code:

https://github.com/jrouwe/JoltPhysics.js/blob/3c4b7f1530b50eb9e377417710d86eda84ec2d03/Examples/js/example.js#L74-L91

For more information on both questions, see: 

https://jrouwe.github.io/JoltPhysics/

Sections 'Broad Phase' and 'Collision filtering'.

If this isn't clear, let me know and I'll try to update the text to make it more clear.

**Comment by LeXXik** [comment]
Thank you for the additional details! After implementing the filtering together with your comments and the docs, it is clear.
